### DISCUSSION TITLE
Custom polygon from Sprite Frame

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [axmolengine/axmol#1974] [General]
Hi,

How can we create a custom polygon physic body from Sprite Frame?

Today the auto polygon only support filename, but by image frame is inside atlas.

Example of what im doing today in this simple code:

```cpp
auto sprite = Sprite::initWithSpriteFrameName("player1");
auto physicsBody = PhysicsBody::createCircle(sprite->getContentSize().width / 2, PhysicsMaterial(1.0f, 0.1f, 1.0f));
sprite->addComponent(physicsBody);
```

The problem is that it don't follow the sprite contours, see:

![Untitled 2](https://github.com/axmolengine/axmol/assets/395096/a19f2af5-1016-4aa3-b623-ed6c475485dd)

I search in some websites for cocos2d-x versions of any code to i port, but i don't find anyone.

Thanks.

**Comment by paulocoutinhox** [comment]
Maybe a new method/constructor here:

https://github.com/axmolengine/axmol/blob/dev/core/2d/AutoPolygon.cpp#L173-L185
and
https://github.com/axmolengine/axmol/blob/dev/core/2d/AutoPolygon.cpp#L725-L729

That support Texture2D/Image.

**Reply by paulocoutinhox** [reply]
Sample code:

```cpp
#include "2d/AutoPolygon.h"
#include "poly2tri/poly2tri.h"
#include "base/Director.h"
#include "base/axstd.h"
#include "renderer/TextureCache.h"
#include "renderer/Texture2D.h"
#include "platform/CCImage.h"
#include "2d/Scale9Sprite.h"
#include "clipper2/clipper.h"
#include <algorithm>
#include <math.h>

// [existing code]

NS_AX_BEGIN

AutoPolygon::AutoPolygon(Texture2D* texture)
    : _image(nullptr), _data(nullptr), _filename(""), _width(0), _height(0), _scaleFactor(0)
{
    AXASSERT(texture != nullptr, "Texture2D is null");

    _image = new Image();
    bool initSuccess = texture->getUIImage(_image);
    AXASSERT(initSuccess, "Failed to create Image from Texture2D");

    AXASSERT(_image->getPixelFormat() == backend::PixelFormat::RGBA8,
             "unsupported format, currently only supports rgba8888");
    
    _data = _image->getData();
    _width = _image->getWidth();
    _height = _image->getHeight();
    _scaleFactor = Director::getInstance()->getContentScaleFactor();
}

AutoPolygon::AutoPolygon(Image* image)
    : _image(image), _data(nullptr), _filename(""), _width(0), _height(0), _scaleFactor(0)
{
    AXASSERT(image != nullptr, "Image is null");
    AXASSERT(_image->getPixelFormat() == backend::PixelFormat::RGBA8,
             "unsupported format, currently only supports rgba8888");

    _data = _image->getData();
    _width = _image->getWidth();
    _height = _image->getHeight();
    _scaleFactor = Director::getInstance()->getContentScaleFactor();
}

AutoPolygon::AutoPolygon(Scale9Sprite* sprite)
    : _image(nullptr), _data(nullptr), _filename(""), _width(0), _height(0), _scaleFactor(0)
{
    AXASSERT(sprite != nullptr, "Scale9Sprite is null");

    auto texture = sprite->getSprite()->getTexture();
    AXASSERT(texture != nullptr, "Scale9Sprite's texture is null");

    _image = new Image();
    bool initSuccess = texture->getUIImage(_image);
    AXASSERT(initSuccess, "Failed to create Image from Texture2D");

    AXASSERT(_image->getPixelFormat() == backend::PixelFormat::RGBA8,
             "unsupported format, currently only supports rgba8888");
    
    _data = _image->getData();
    _width = _image->getWidth();
    _height = _image->getHeight();
    _scaleFactor = Director::getInstance()->getContentScaleFactor();
}

// [existing code]

PolygonInfo AutoPolygon::generatePolygon(Texture2D* texture, const Rect& rect, float epsilon, float threshold)
{
    AutoPolygon ap(texture);
    return ap.generateTriangles(rect, epsilon, threshold);
}

PolygonInfo AutoPolygon::generatePolygon(Image* image, const Rect& rect, float epsilon, float threshold)
{
    AutoPolygon ap(image);
    return ap.generateTriangles(rect, epsilon, threshold);
}

PolygonInfo AutoPolygon::generatePolygon(Scale9Sprite* sprite, const Rect& rect, float epsilon, float threshold)
{
    AutoPolygon ap(sprite);
    return ap.generateTriangles(rect, epsilon, threshold);
}

// [existing code]

NS_AX_END
```

**Comment by smilediver** [comment]
Can't you get `PolygonInfo` from `SpriteFrame`?

**Reply by paulocoutinhox** [reply]
It always return 0:

<img width="634" alt="image" src="https://github.com/axmolengine/axmol/assets/395096/1714af58-a4fc-4705-ac19-01eb5688ddfb">

<img width="650" alt="image" src="https://github.com/axmolengine/axmol/assets/395096/84452205-0fe5-4c62-96de-f47e2b21a813">

**Reply by paulocoutinhox** [reply]
The format of plist sprite frame need have the vertices info to be loaded in this case by the spritesheet generator tool:

https://www.codeandweb.com/knowledgebase/cocos2d-plist-file-format#polygon-mesh-data

I don't know if any tool do this, but it is not calculated in real time, but is retrieved from plist frame data.

My suggestion is a way to do from our code in realtime.

Also, this kind of info inside the link will be nice have in our wiki here.

**Reply by paulocoutinhox** [reply]
If anyone use TexturePacker, it can export this kind of info if your select it when export, but i don't have license for that tool:
https://www.codeandweb.com/texturepacker/documentation/custom-exporter#sprite-type

**Comment by GamesStudioTwo** [comment]
Made a small video for you hope it helps.

https://www.youtube.com/watch?v=qDe_e-sAXU0

**Reply by paulocoutinhox** [reply]
Hum....very nice man...so it don't solve my problem :(

Im trying creating here a converter from Texture2D to Image to update AutoPolygon class. If we convert from texture to image, we can use AutoPolygon class.

If i reach, i will post here the results.

**Reply by paulocoutinhox** [reply]
I think it will not be possible, to retrive Image object from Texture2D because it don't have all pixel data info.

**Reply by smilediver** [reply]
Maybe try loading the image like this: `AutoPolygon(texture->getPath())`. Then use `SpriteFrame`'s rect to create polygons.

**Reply by paulocoutinhox** [reply]
Hi, the SpriteFrame came from Atlas image, so the SpriteFrame is a small piece inside it. In this getPath maybe return empty or the full Atlas image. Example:

![image](https://github.com/axmolengine/axmol/assets/395096/40bb8d4c-8605-432a-84b2-b7d4a9c3424d)

**Reply by smilediver** [reply]
Maybe we misunderstand each other, cause I don't understand the problem... IIRC `texture->getPath()` will always have a path for a texture loaded from a file. So from a given `SpriteFrame` you can extract what texture and image file it was loaded from, and pass that file to `AutoPolygon` to work on that image. Then you can get the rect from `SpriteFrame` to find the location where that sprite is placed in the atlas image, and then pass that rect to `AutoPolygon::generateTriangles()` so it would build the `PolygonInfo` for that sprite in the region limited by rect.

**Comment by aismann** [comment]
> Hi,
> 
> How can we create a custom polygon physic body from Sprite Frame?
> 
> Today the auto polygon only support filename, but by image frame is inside atlas.
> 
> Example of what im doing today in this simple code:
> 
> ```c++
> auto sprite = Sprite::initWithSpriteFrameName("player1");
> auto physicsBody = PhysicsBody::createCircle(sprite->getContentSize().width / 2, PhysicsMaterial(1.0f, 0.1f, 1.0f));
> sprite->addComponent(physicsBody);
> ```
> 
> The problem is that it don't follow the sprite contours, see:
> 
>

Dont use `PhysicsBody::createCircle` in all cases:
Select the right one:
![image](https://github.com/axmolengine/axmol/assets/8652787/289ce47f-1236-4c99-ae83-86400d3a98fe)

**Reply by paulocoutinhox** [reply]
I know, im using this because it fit my needs. Thanks.
